File: src/program/hybrid_access/base/buffer.c
#include <stdlib.h>
#include "buffer.h"

// Initialize the delay buffer
struct buffer* buffer_new() {
    struct buffer *buffer = (struct buffer*) malloc(sizeof(struct buffer));
    if (buffer) {
        buffer->read = 0;
        buffer->write = 0;
    }
    return buffer;
}

// Check if the buffer is empty
inline int buffer_is_empty(struct buffer *buffer) {
    return buffer->read == buffer->write;
}

// Check if the buffer is full
inline int buffer_is_full(struct buffer *buffer) {
    return ((buffer->write + 1) & (BUFFER_SIZE - 1)) == buffer->read;
}

// Enqueue a timed packet to the buffer
// not full check needs to be done
int buffer_enqueue(struct buffer *buffer, struct packet *pkt) {
    if (buffer_is_full(buffer)) {
        return 0;
    }
    buffer->packets[buffer->write] = pkt;

    buffer->write = (buffer->write + 1) & (BUFFER_SIZE - 1);
    return 1;
}

// Dequeue a timed packet from the buffer
// not empty check needs to be done
struct packet* buffer_dequeue(struct buffer *buffer) {
    // assert not empty
    struct packet *pkt = buffer->packets[buffer->read];
    buffer->read = (buffer->read + 1) & (BUFFER_SIZE - 1);
    return pkt;
}

// Amount of readable packets
int buffer_size(struct buffer *buffer) {
    if (buffer->write >= buffer->read) {
        return buffer->write - buffer->read;
    }
    return buffer->write + BUFFER_SIZE - buffer->read;
}

// Destroy the delay buffer and free its memory
void buffer_free(struct buffer *buffer) {
    if (buffer) {
        free(buffer);
    }
}

File: src/program/hybrid_access/base/buffer.lua
module(..., package.seeall)

local ffi = require("ffi")
local bit = require("bit")

require("core.packet_h")

local band = bit.band

PacketBuffer = {}

function PacketBuffer:new(size)
    assert(band(size, (size - 1)) == 0, "size must be a power of two")
    local buf = ffi.new("struct packet *[?]", size)
    local o = {
        buffer = buf,
        read = 0,
        write = 0,
        max = size
    }
    setmetatable(o, self)
    self.__index = self
    return o
end

function PacketBuffer:size()
    if self.write >= self.read then
        return self.write - self.read
    end
    return self.write + self.max - self.read
end

function PacketBuffer:is_full()
    return band(self.write + 1, self.max - 1) == self.read;
end

function PacketBuffer:dequeue()
    -- assert not empty
    local pkt = self.buf[self.read]
    self.read = band(self.read + 1, self.max - 1)
    return pkt;
end

function PacketBuffer:enqueue(pkt)
    if self:is_full() then
        return 0
    end
    self.buffer[self.write] = pkt;
    self.write = band(self.write, self.max - 1)

    return 1;
end